nbsphinx_link/__init__.py

```diff
@@ -19,53 +19,91 @@
 Further keys might be added in the future.
 """
 
-import errno
 import json
 import os
 import shutil
 
 from docutils import io, nodes, utils
 from docutils.utils.error_reporting import SafeString, ErrorString
-from docutils.nodes import document
+import docutils  # noqa: F401
 from nbsphinx import NotebookParser, NotebookError, _ipynbversion
 import nbformat
 from sphinx.util.logging import getLogger
 
 from ._version import __version__
 
 
-def copy_directory(src, dest):
+def register_dependency(file_path, document):
     """
-    Copies a directory or file from the path ``src`` to ``dest``.
-    Originally from: https://www.pythoncentral.io/how-to-recursively-copy-a-directory-folder-in-python/
-    Which is licensed under the MIT License: https://opensource.org/licenses/MIT
+    Registers files as dependency, so sphinx rebuilds the docs
+    when they changed.
 
+    Parameters
+    ----------
+    file_path : str
+        [description]
+    document: docutils.nodes.document
+        Parsed document instance.
+    """
+    document.settings.record_dependencies.add(file_path)
+    document.settings.env.note_dependency(file_path)
+
+
+def copy_file(src, dest):
+    """
+    Copies a singe file from ``src`` to ``dest``.
+
+    Parameters
+    ----------
+    src : str
+        Path to the source file.
+    dest : str
+        Path to the destination file.
+    """
+    logger = getLogger(__name__)
+    try:
+        shutil.copy(src, dest)
+    except (OSError) as e:
+        logger.warning(
+            "The the file {} couldn't be copied. "
+            "Error:\n {}".format(src, e)
+        )
+
+
+def copy_and_register_files(src, dest, document):
+    """
+    Copies a directory or file from the path ``src`` to ``dest``
+    and registers all files as dependency,
+    so sphinx rebuilds the docs when they changed.
 
     Parameters
     ----------
     src : str
         Path to the source directory or file
     dest : str
         Path to the destination directory or file
+    document: docutils.nodes.document
+        Parsed document instance.
     """
-    try:
-        # ensure that files will be overwritten
-        shutil.rmtree(dest, ignore_errors=True)
-        shutil.copytree(src, dest)
-    except OSError as e:
-        # If the error was caused because the source wasn't a directory
-        if e.errno == errno.ENOTDIR:
-            shutil.copy(src, dest)
-        else:
-            print('Directory not copied. Error: {}'.format(e))
+    if os.path.isdir(src):
+        for root, _, filenames in os.walk(src):
+            for filename in filenames:
+                src_path = os.path.abspath(os.path.join(root, filename))
+                if not os.path.exists(dest):
+                    os.makedirs(dest)
```

Comment:
This existence check seems like it does not belong here. It should be moved outside the filename loop, and should check for and create `os.path.join(dest, os.relpath(root, src))` (I think, probably needs tweaking). You could also just call os.makedirs with `exist_ok=True`.

Reply:
I moved it out if the loop.
``exist_ok=True`` would break the py2.7 compatibility, since it was [added in python 3.2](https://docs.python.org/3.7/library/os.html#os.makedirs) .

Reply:
Ok on the py2 bit. I pushed a commit to ensure that this code handles "src" having subdirectories. Do you agree with the changes?

Reply:
Thats nice, i will also add an example for that in the notebook.

Reply:
I made ``dst_root`` also the target to copy files.